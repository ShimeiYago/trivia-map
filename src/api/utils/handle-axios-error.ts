import { AxiosError } from 'axios';

export function handleAxiosError(axiosError: AxiosError): ApiError {
  if (axiosError.response) {
    // when API is active but response includes error
    const error: ApiError = {
      status: axiosError.response.status,
      data: axiosError.response.data,
      errorMsg: axiosError.message,
    };
    return error;
  } else {
    // when API is not active
    const error: ApiError = {
      status: 500,
      data: {},
      errorMsg: axiosError.message,
    };
    return error;
  }
}

export type ApiError = {
  status: number;
  data: unknown;
  errorMsg: string;
};
