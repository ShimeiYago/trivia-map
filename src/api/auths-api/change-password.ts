import { BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function changePassword(
  password1: string,
  password2: string,
): Promise<void> {
  const axiosInstance = getAxiosInstance({}, {});

  const requestData: ChangePasswordRequest = {
    password1: password1,
    password2: password2,
  };

  try {
    // TODO: Set reasonable timeout
    await axiosInstance.post(`${BASE_URL}/auths/password/change`, requestData);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type ChangePasswordRequest = {
  password1: string;
  password2: string;
};

export type ValidationError = {
  password1?: string[];
  password2?: string[];
};
