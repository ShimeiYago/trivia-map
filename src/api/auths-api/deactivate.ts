import { API_TIMEOUT, BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function deactivate(): Promise<void> {
  const axiosInstance = getAxiosInstance({ timeout: API_TIMEOUT.long }, {});

  try {
    await axiosInstance.put(`${BASE_URL}/auths/deactivate/`);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}
