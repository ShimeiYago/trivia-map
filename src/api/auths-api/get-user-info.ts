import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { User } from 'types/user';
import { mockGetUserInfoResponse } from 'api/mock/auths-response/get-user-info';

export async function getUserInfo(): Promise<User> {
  const axiosInstance = getAxiosInstance({}, mockGetUserInfoResponse);

  try {
    // TODO: Set reasonable timeout
    const res: AxiosResponse<User> = await axiosInstance.get(`${BASE_URL}/auths/user`);
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}
