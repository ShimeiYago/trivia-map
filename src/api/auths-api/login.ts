import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { mockLoginResponse } from 'api/mock/auths-response/login';
import { User } from 'types/user';

export async function login(
  email: string,
  password: string,
): Promise<LoginResponse> {
  const axiosInstance = getAxiosInstance({}, mockLoginResponse);

  const requestData: LoginRequest = {
    email: email,
    password: password,
  };

  try {
    // TODO: Set reasonable timeout
    const res: AxiosResponse<LoginResponse> = await axiosInstance.post(
      `${BASE_URL}/auths/login`,
      requestData,
    );
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type LoginRequest = {
  email: string;
  password: string;
};

export type LoginResponse = {
  access_token: string;
  refresh_token: string;
  user: User;
};

export type ValidationError = {
  email?: string[];
  password?: string[];
  non_field_errors?: string[];
};
