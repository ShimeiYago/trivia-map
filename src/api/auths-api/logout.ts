import { BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function logout(): Promise<void> {
  const axiosInstance = getAxiosInstance({}, {});

  try {
    await axiosInstance.post(`${BASE_URL}/auths/logout`);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}
