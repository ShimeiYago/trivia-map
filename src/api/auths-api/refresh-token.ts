import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { mockRefreshTokenResponse } from 'api/mock/auths-response/refresh-token';

export async function refreshToken(): Promise<RefreshTokenResponse> {
  const axiosInstance = getAxiosInstance({}, mockRefreshTokenResponse);

  try {
    const res: AxiosResponse<RefreshTokenResponse> = await axiosInstance.post(
      `${BASE_URL}/auths/token/refresh`,
    );
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}

export type RefreshTokenResponse = {
  access: string;
  access_token_expiration: string;
};
