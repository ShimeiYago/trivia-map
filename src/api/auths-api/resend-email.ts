import { BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function resendEmail(email: string): Promise<void> {
  const axiosInstance = getAxiosInstance({}, {});

  const requestData: VerifyEmailRequest = {
    email: email,
  };

  try {
    // TODO: Set reasonable timeout
    await axiosInstance.post(`${BASE_URL}/auths/registration/resend-email`, requestData);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type VerifyEmailRequest = {
  email: string;
};

export type ValidationError = {
  email?: string[];
  non_field_errors?: string[];
};
