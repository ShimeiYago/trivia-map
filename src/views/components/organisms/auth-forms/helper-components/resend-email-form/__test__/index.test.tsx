import { shallow, ShallowWrapper } from 'enzyme';
import { ResendEmailForm, Props, State } from '..';
import * as ResendEmailModule from 'api/auths-api/resend-email';
import { ApiError } from 'api/utils/handle-axios-error';

let wrapper: ShallowWrapper<Props, State, ResendEmailForm>;

let loginSpy: jest.SpyInstance;

const basicProps: Props = {
  switchMode: jest.fn(),
  onChangeEmail: jest.fn(),
  email: '',
};

describe('Shallow Snapshot Tests', () => {
  beforeEach(() => {
    wrapper = shallow(<ResendEmailForm {...basicProps} />);
  });

  it('basic', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('error case', () => {
    wrapper.setState({
      errorTitle: 'error title',
      errorMessages: ['error1', 'error2'],
      formError: {
        email: ['email is invalid'],
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('success case', () => {
    wrapper.setState({
      localLoadingState: 'success',
    });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('handleSubmit', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    loginSpy = jest.spyOn(ResendEmailModule, 'resendEmail');
  });

  it('should set localLoadingState success when api succeed', async () => {
    loginSpy.mockResolvedValue({});

    wrapper = shallow(<ResendEmailForm {...basicProps} />);
    const instance = wrapper.instance();

    await instance['handleSubmit']();

    expect(instance.state.localLoadingState).toBe('success');
  });

  it('should set form error when api have validation error', async () => {
    const apiError: ApiError<ResendEmailModule.ValidationError> = {
      status: 400,
      data: {
        email: ['email is invalid'],
        non_field_errors: ['unknown error'],
      },
      errorMsg: '400 request is invalid',
    };
    loginSpy.mockRejectedValue(apiError);

    wrapper = shallow(<ResendEmailForm {...basicProps} />);
    const instance = wrapper.instance();

    await instance['handleSubmit']();

    expect(instance.state.formError).toEqual({
      email: ['email is invalid'],
    });
  });

  it('should set localLoadingState error when api fail', async () => {
    loginSpy.mockRejectedValue(new Error());

    wrapper = shallow(<ResendEmailForm {...basicProps} />);
    const instance = wrapper.instance();

    await instance['handleSubmit']();

    expect(instance.state.localLoadingState).toBe('error');
  });
});
