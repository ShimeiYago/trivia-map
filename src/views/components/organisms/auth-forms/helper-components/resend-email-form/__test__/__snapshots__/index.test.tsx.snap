// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Snapshot Tests basic 1`] = `
<ForwardRef(Stack)
  spacing={1}
  sx={
    Object {
      "px": 1,
      "py": 2,
    }
  }
>
  <ForwardRef(Typography)
    align="center"
    component="h1"
    variant="h5"
  >
    確認用メール再送
  </ForwardRef(Typography)>
  <ForwardRef(Typography)
    component="p"
    variant="subtitle2"
  >
    アカウント作成時に送信した確認メールをもう一度送信します。
  </ForwardRef(Typography)>
  <ForwardRef(Box)
    component="form"
    noValidate={true}
  >
    <ForwardRef(TextField)
      autoComplete="email"
      disabled={false}
      error={false}
      fullWidth={true}
      id="email"
      label="メールアドレス"
      margin="normal"
      name="email"
      onChange={[MockFunction]}
      required={true}
      value=""
    />
    <ForwardRef(LoadingButton)
      disabled={false}
      fullWidth={true}
      loading={false}
      onClick={[Function]}
      sx={
        Object {
          "mb": 2,
          "mt": 3,
        }
      }
      variant="contained"
    >
      メールを再送信する
    </ForwardRef(LoadingButton)>
    <ForwardRef(Grid)
      container={true}
    >
      <ForwardRef(Grid)
        item={true}
        xs={true}
      >
        <ForwardRef(Button)
          variant="text"
        >
          ログイン
        </ForwardRef(Button)>
      </ForwardRef(Grid)>
      <ForwardRef(Grid)
        item={true}
      >
        <ForwardRef(Button)
          variant="text"
        >
          アカウント作成
        </ForwardRef(Button)>
      </ForwardRef(Grid)>
    </ForwardRef(Grid)>
  </ForwardRef(Box)>
</ForwardRef(Stack)>
`;

exports[`Shallow Snapshot Tests error case 1`] = `
<ForwardRef(Stack)
  spacing={1}
  sx={
    Object {
      "px": 1,
      "py": 2,
    }
  }
>
  <ForwardRef(Typography)
    align="center"
    component="h1"
    variant="h5"
  >
    確認用メール再送
  </ForwardRef(Typography)>
  <ForwardRef(Typography)
    component="p"
    variant="subtitle2"
  >
    アカウント作成時に送信した確認メールをもう一度送信します。
  </ForwardRef(Typography)>
  <ForwardRef(Box)>
    <HeaderErrorMessages
      errorMessages={
        Array [
          "error1",
          "error2",
        ]
      }
      errorTitle="error title"
    />
  </ForwardRef(Box)>
  <ForwardRef(Box)
    component="form"
    noValidate={true}
  >
    <ForwardRef(TextField)
      autoComplete="email"
      disabled={false}
      error={true}
      fullWidth={true}
      helperText={
        Array [
          "email is invalid",
        ]
      }
      id="email"
      label="メールアドレス"
      margin="normal"
      name="email"
      onChange={[MockFunction]}
      required={true}
      value=""
    />
    <ForwardRef(LoadingButton)
      disabled={false}
      fullWidth={true}
      loading={false}
      onClick={[Function]}
      sx={
        Object {
          "mb": 2,
          "mt": 3,
        }
      }
      variant="contained"
    >
      メールを再送信する
    </ForwardRef(LoadingButton)>
    <ForwardRef(Grid)
      container={true}
    >
      <ForwardRef(Grid)
        item={true}
        xs={true}
      >
        <ForwardRef(Button)
          variant="text"
        >
          ログイン
        </ForwardRef(Button)>
      </ForwardRef(Grid)>
      <ForwardRef(Grid)
        item={true}
      >
        <ForwardRef(Button)
          variant="text"
        >
          アカウント作成
        </ForwardRef(Button)>
      </ForwardRef(Grid)>
    </ForwardRef(Grid)>
  </ForwardRef(Box)>
</ForwardRef(Stack)>
`;

exports[`Shallow Snapshot Tests success case 1`] = `
<ForwardRef(Stack)
  spacing={1}
  sx={
    Object {
      "px": 1,
      "py": 2,
    }
  }
>
  <ForwardRef(Typography)
    align="center"
    component="h1"
    variant="h5"
  >
    確認用メール再送
  </ForwardRef(Typography)>
  <ForwardRef(Typography)
    component="p"
    variant="subtitle2"
  >
    アカウント作成時に送信した確認メールをもう一度送信します。
  </ForwardRef(Typography)>
  <ForwardRef(Alert)>
    入力されたメールアドレス宛に確認用メールを再送信しました。メールの内容に従ってアカウント作成を完了してください。
  </ForwardRef(Alert)>
  <ForwardRef(Box)
    component="form"
    noValidate={true}
  >
    <ForwardRef(TextField)
      autoComplete="email"
      disabled={false}
      error={false}
      fullWidth={true}
      id="email"
      label="メールアドレス"
      margin="normal"
      name="email"
      onChange={[MockFunction]}
      required={true}
      value=""
    />
    <ForwardRef(LoadingButton)
      disabled={false}
      fullWidth={true}
      loading={false}
      onClick={[Function]}
      sx={
        Object {
          "mb": 2,
          "mt": 3,
        }
      }
      variant="contained"
    >
      メールを再送信する
    </ForwardRef(LoadingButton)>
    <ForwardRef(Grid)
      container={true}
    >
      <ForwardRef(Grid)
        item={true}
        xs={true}
      >
        <ForwardRef(Button)
          variant="text"
        >
          ログイン
        </ForwardRef(Button)>
      </ForwardRef(Grid)>
      <ForwardRef(Grid)
        item={true}
      >
        <ForwardRef(Button)
          variant="text"
        >
          アカウント作成
        </ForwardRef(Button)>
      </ForwardRef(Grid)>
    </ForwardRef(Grid)>
  </ForwardRef(Box)>
</ForwardRef(Stack)>
`;
